src/Dependencies/PooledObjects/PooledDelegates.cs

```diff
@@ -0,0 +1,404 @@
+            => GetPooledDelegate<FuncWithBoundArgument<T1, T2, TArg, TResult>, TArg, Func<T1, T2, TArg, TResult>, Func<T1, T2, TResult>>(unboundFunction, argument, out boundFunction);
+
+        /// <summary>
+        /// Gets a <see cref="Func{T1, T2, T3, TResult}"/> delegate, which calls <paramref name="unboundFunction"/> with the
+        /// specified <paramref name="argument"/>. The resulting <paramref name="boundFunction"/> may be called any
+        /// number of times until the returned <see cref="Releaser"/> is disposed.
+        /// </summary>
+        /// <example>
+        /// <para>The following example shows the use of a capturing delegate for a predicate that requires an
+        /// argument:</para>
+        ///
+        /// <code>
+        /// int x = 3;
+        /// RunWithPredicate((a, b, c) => this.IsSomething(a, b, c, x));
+        /// </code>
+        ///
+        /// <para>The following example shows the use of a pooled delegate to avoid capturing allocations for the same
+        /// predicate:</para>
+        ///
+        /// <code>
+        /// int x = 3;
+        /// using var _ = GetPooledFunction((a, b, c, arg) => arg.self.IsSomething(a, b, c, arg.x), (self: this, x), out Func&lt;int, int, int, bool&gt; predicate);
+        /// RunWithPredicate(predicate);
+        /// </code>
+        /// </example>
+        /// <typeparam name="T1">The type of the first parameter of the bound function.</typeparam>
+        /// <typeparam name="T2">The type of the second parameter of the bound function.</typeparam>
+        /// <typeparam name="T3">The type of the third parameter of the bound function.</typeparam>
+        /// <typeparam name="TArg">The type of argument to pass to <paramref name="unboundFunction"/>.</typeparam>
+        /// <typeparam name="TResult">The type of the return value of the function.</typeparam>
+        /// <param name="unboundFunction">The unbound function delegate.</param>
+        /// <param name="argument">The argument to pass to the unbound function delegate.</param>
+        /// <param name="boundFunction">A delegate which calls <paramref name="unboundFunction"/> with the specified
+        /// <paramref name="argument"/>.</param>
+        /// <returns>A disposable <see cref="Releaser"/> which returns the object to the delegate pool.</returns>
+        public static Releaser GetPooledFunction<T1, T2, T3, TArg, TResult>(Func<T1, T2, T3, TArg, TResult> unboundFunction, TArg argument, out Func<T1, T2, T3, TResult> boundFunction)
+            => GetPooledDelegate<FuncWithBoundArgument<T1, T2, T3, TArg, TResult>, TArg, Func<T1, T2, T3, TArg, TResult>, Func<T1, T2, T3, TResult>>(unboundFunction, argument, out boundFunction);
+
+        public struct Releaser : IDisposable
+        {
+            private readonly object _pooledObject;
+            private readonly Action<object> _releaser;
+
+            internal Releaser(object pooledObject, Action<object> releaser)
```

Comment:
>object [](start = 30, length = 6)

```AbstractDelegateWithBoundArgument```?

Reply:
Not possible without making `Releaser` generic.

Reply:
Edit: This is now getting covered by https://github.com/dotnet/roslyn/pull/40685#discussion_r362630244